### PR TITLE
Feature/exam service

### PR DESCRIPTION
**Pull request by david-buderus**
### Reason for this PR
- Exams and exam results should get implemented

### Changes in this PR
- Added ExamService
- Added ExamResultService
- Added support in AdmissionService to get a ExamAdmissionProposal
- Allowed students to fetch their own username, given the enrollmentId in the CertificateService
- Added validation information of Exam related objects to ValidationConfiguration, like ExamAdmission etc. in ConfigurationService
- Added grades and exam types to Configuration in ConfigurationService

- [x] Test written
- [x] Changelog updated

### CODE REVIEW

File: product_code/admission_service/api/src/main/scala/de/upb/cs/uc4/admission/api/AdmissionService.scala [hunk 1]
```diff
@@ -22,12 +22,14 @@ trait AdmissionService extends UC4HyperledgerService {
...
-  def getProposalAddCourseAdmission: ServiceCall[CourseAdmission, UnsignedProposal]
+  /** Returns exam admissions */
+  def getExamAdmissions(username: Option[String], admissionIds: Option[String], examIds: Option[String]): ServiceCall[NotUsed, Seq[ExamAdmission]]
 
-  /** Gets a proposal for dropping a course admission */
-  def getProposalDropCourseAdmission: ServiceCall[DropAdmission, UnsignedProposal]
+  /** Gets a proposal for adding a exam admission */
+  def getProposalAddAdmission: ServiceCall[AbstractAdmission, UnsignedProposal]
 
+  /** Gets a proposal for dropping a course admission */
```
Comment (ldklenner) [inline product_code/admission_service/api/src/main/scala/de/upb/cs/uc4/admission/api/AdmissionService.scala]:
```suggestion
  /** Gets a proposal for dropping an admission */
```

File: product_code/admission_service/api/src/main/scala/de/upb/cs/uc4/admission/api/AdmissionService.scala [hunk 2]
```diff
@@ -22,12 +22,14 @@ trait AdmissionService extends UC4HyperledgerService {
...
   def getCourseAdmissions(username: Option[String], courseId: Option[String], moduleId: Option[String]): ServiceCall[NotUsed, Seq[CourseAdmission]]
 
-  /** Gets a proposal for adding a course admission */
-  def getProposalAddCourseAdmission: ServiceCall[CourseAdmission, UnsignedProposal]
+  /** Returns exam admissions */
+  def getExamAdmissions(username: Option[String], admissionIds: Option[String], examIds: Option[String]): ServiceCall[NotUsed, Seq[ExamAdmission]]
 
-  /** Gets a proposal for dropping a course admission */
-  def getProposalDropCourseAdmission: ServiceCall[DropAdmission, UnsignedProposal]
+  /** Gets a proposal for adding a exam admission */
```
Comment (ldklenner) [inline product_code/admission_service/api/src/main/scala/de/upb/cs/uc4/admission/api/AdmissionService.scala]:
```suggestion
  /** Gets a proposal for adding an admission */
```

File: product_code/admission_service/api/src/test/scala/de/upb/cs/uc4/admission/AdmissionServiceStub.scala [hunk 3]
```diff
@@ -3,35 +3,46 @@ package de.upb.cs.uc4.admission
...
   }
 
-  /** Gets a proposal for adding a course admission */
-  override def getProposalAddCourseAdmission: ServiceCall[CourseAdmission, UnsignedProposal] = ServiceCall {
+  /** Gets a proposal for adding a exam admission */
+  override def getProposalAddAdmission: ServiceCall[AbstractAdmission, UnsignedProposal] = ServiceCall {
     _ => Future.successful(UnsignedProposal(""))
   }
 
   /** Gets a proposal for dropping a course admission */
```
Comment (ldklenner) [inline product_code/admission_service/api/src/test/scala/de/upb/cs/uc4/admission/AdmissionServiceStub.scala]:
```suggestion
  /** Gets a proposal for dropping an admission */
```

File: product_code/admission_service/api/src/test/scala/de/upb/cs/uc4/admission/AdmissionServiceStub.scala [hunk 4]
```diff
@@ -3,35 +3,46 @@ package de.upb.cs.uc4.admission
...
+  }
+
+  /** Returns exam admissions */
+  override def getExamAdmissions(username: Option[String], admissionIds: Option[String], examIds: Option[String]): ServiceCall[NotUsed, Seq[ExamAdmission]] = ServiceCall {
+    _ => Future.successful(examAdmissions)
   }
 
-  /** Gets a proposal for adding a course admission */
-  override def getProposalAddCourseAdmission: ServiceCall[CourseAdmission, UnsignedProposal] = ServiceCall {
+  /** Gets a proposal for adding a exam admission */
```
Comment (ldklenner) [inline product_code/admission_service/api/src/test/scala/de/upb/cs/uc4/admission/AdmissionServiceStub.scala]:
```suggestion
  /** Gets a proposal for adding an admission */
```

File: product_code/course_service/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,9 @@
+# [v0.17.1](https://github.com/upb-uc4/University-Credits-4.0/compare/course-v0.16.1...course-v0.17.1) (2021-02-12)
+## Feature
+## Refactor
+ - changed validation timeout from 500ms to 5000ms
```
Comment (ldklenner) [inline product_code/course_service/CHANGELOG.md]:
```suggestion
 - Changed validation timeout from 500ms to 5000ms
```

File: product_code/exam_result_service/CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# [WIP - v0.18.1](https://github.com/upb-uc4/University-Credits-4.0/compare/examresult-v0.18.1...examresult-v0.18.1) (2021-XX-XX)
+## Feature
+ - Added exam result service with endpoints for
```
Comment (ldklenner) [inline product_code/exam_result_service/CHANGELOG.md:3]:
```suggestion
 - Added exam result service with endpoints for
  - getting all exam result entries
  - fetching a proposal for creating results for the exam
```

File: product_code/exam_result_service/impl/src/main/scala/de/upb/cs/uc4/examresult/impl/ExamResultServiceImpl.scala [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+import de.upb.cs.uc4.operation.model.JsonOperationId
+import de.upb.cs.uc4.shared.client.exceptions.{ DetailedError, ErrorType, SimpleError, UC4Exception, UC4NonCriticalException }
+import de.upb.cs.uc4.shared.server.ServiceCallFactory._
+import org.slf4j.{ Logger, LoggerFactory }
+import play.api.Environment
+
+import scala.concurrent.{ Await, ExecutionContext, Future, TimeoutException }
+import scala.concurrent.duration._
+
+/** Implementation of the ExamService */
```
Comment (ldklenner) [inline product_code/exam_result_service/impl/src/main/scala/de/upb/cs/uc4/examresult/impl/ExamResultServiceImpl.scala]:
```suggestion
/** Implementation of the ExamResultService */
```

File: product_code/exam_result_service/impl/src/main/scala/de/upb/cs/uc4/examresult/impl/actor/ExamResultBehaviour.scala [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+      )
+  }
+
+  /** The companion object */
+  override val companionObject: HyperledgerActorObject = ExamResultBehaviour
+}
+
+object ExamResultBehaviour extends HyperledgerActorObject {
+  /** The EntityTypeKey of this actor */
+  override val typeKey: EntityTypeKey[HyperledgerBaseCommand] = EntityTypeKey[HyperledgerBaseCommand]("uc4exam")
```
Comment (ldklenner) [inline product_code/exam_result_service/impl/src/main/scala/de/upb/cs/uc4/examresult/impl/actor/ExamResultBehaviour.scala]: Is it intended that the TypeKey has the same ID as the `ExamBehaviour` TypeKey?

File: product_code/exam_service/CHANGELOG.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# [WIP - v0.18.1](https://github.com/upb-uc4/University-Credits-4.0/compare/exam-v0.18.1...exam-v0.18.1) (2021-XX-XX)
+## Feature
+ - Added exam service with endpoints for
```
Comment (ldklenner) [inline product_code/exam_service/CHANGELOG.md:3]:
```suggestion
 - Added exam service with endpoints for
   - getting all exams
  - fetching a proposal for creating the exam
```

File: product_code/exam_service/api/src/main/scala/de/upb/cs/uc4/exam/api/ExamService.scala [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+import akka.{ Done, NotUsed }
+import com.lightbend.lagom.scaladsl.api.deser.MessageSerializer
+import com.lightbend.lagom.scaladsl.api.transport.Method
+import com.lightbend.lagom.scaladsl.api.{ Descriptor, Service, ServiceCall }
+import de.upb.cs.uc4.exam.model.Exam
+import de.upb.cs.uc4.hyperledger.api.UC4HyperledgerService
+import de.upb.cs.uc4.hyperledger.api.model.UnsignedProposal
+import de.upb.cs.uc4.shared.client.message_serialization.CustomMessageSerializer
+
+/** The MatriculationService interface.
```
Comment (ldklenner) [inline product_code/exam_service/api/src/main/scala/de/upb/cs/uc4/exam/api/ExamService.scala]:
```suggestion
/** The ExamService interface.
```

File: product_code/exam_service/api/src/main/scala/de/upb/cs/uc4/exam/api/ExamService.scala [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  *
+  * This describes everything that Lagom needs to know about how to serve and
+  * consume the MatriculationService.
```
Comment (ldklenner) [inline product_code/exam_service/api/src/main/scala/de/upb/cs/uc4/exam/api/ExamService.scala]:
```suggestion
  * consume the ExamService.
```

File: product_code/exam_service/api/src/test/scala/de/upb/cs/uc4/exam/ExamServiceStub.scala [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+import akka.{ Done, NotUsed }
+import com.lightbend.lagom.scaladsl.api.ServiceCall
+import de.upb.cs.uc4.exam.api.ExamService
+import de.upb.cs.uc4.exam.model.Exam
+import de.upb.cs.uc4.hyperledger.api.model.{ JsonHyperledgerVersion, UnsignedProposal }
+import de.upb.cs.uc4.shared.client.JsonUtility.ToJsonUtil
+
+import scala.concurrent.Future
+
+class ExamServiceStub() extends ExamService with DefaultTestExams {
```
Comment (ldklenner) [inline product_code/exam_service/api/src/test/scala/de/upb/cs/uc4/exam/ExamServiceStub.scala]:
```suggestion
class ExamServiceStub extends ExamService with DefaultTestExams {
```

File: product_code/exam_result_service/impl/src/main/scala/de/upb/cs/uc4/examresult/impl/actor/ExamResultBehaviour.scala (hunk 8 above)
Comment (david-buderus) [reply product_code/exam_result_service/impl/src/main/scala/de/upb/cs/uc4/examresult/impl/actor/ExamResultBehaviour.scala]: No, fixed it

**Review by RBeierling — APPROVED** [review]

File: product_code/certificate_service/CHANGELOG.md [hunk 13]
```diff
@@ -1,3 +1,15 @@
+# [v0.18.1](https://github.com/upb-uc4/University-Credits-4.0/compare/certificate-v0.16.2...certificate-v0.18.2) (2021-02-18)
```
Comment (dhelmert) [inline product_code/certificate_service/CHANGELOG.md]:
```suggestion
# [v0.18.1](https://github.com/upb-uc4/University-Credits-4.0/compare/certificate-v0.16.2...certificate-v0.18.1) (2021-02-18)
```

**Review by lrschaef — APPROVED** [review]
Changelogs need to be updated on develop.

**Review by ldklenner — APPROVED** [review]

File: product_code/exam_result_service/api/src/test/scala/de/upb/cs/uc4/examresult/DefaultTestExamResultEntries.scala [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+  var exam0ResultEntry2: ExamResultEntry = ExamResultEntry(student2.username + "enrollmentId", exam0.examId, ConfigurationCollection.grades.apply(2))
+
+  var exam1ResultEntry0: ExamResultEntry = ExamResultEntry(student0.username + "enrollmentId", exam1.examId, ConfigurationCollection.grades.head)
+  var exam1ResultEntry1: ExamResultEntry = ExamResultEntry(student1.username + "enrollmentId", exam1.examId, ConfigurationCollection.grades.apply(1))
+  var exam1ResultEntry2: ExamResultEntry = ExamResultEntry(student2.username + "enrollmentId", exam1.examId, ConfigurationCollection.grades.apply(2))
+
+  var exam3ResultEntry0: ExamResultEntry = ExamResultEntry(student0.username + "enrollmentId", exam3.examId, ConfigurationCollection.grades.head)
+  var exam3ResultEntry1: ExamResultEntry = ExamResultEntry(student1.username + "enrollmentId", exam3.examId, ConfigurationCollection.grades.apply(1))
+  var exam3ResultEntry2: ExamResultEntry = ExamResultEntry(student2.username + "enrollmentId", exam3.examId, ConfigurationCollection.grades.apply(2))
```
Comment (dhelmert) [inline product_code/exam_result_service/api/src/test/scala/de/upb/cs/uc4/examresult/DefaultTestExamResultEntries.scala]:
```suggestion
  var exam0ResultEntry1: ExamResultEntry = ExamResultEntry(student1.username + "enrollmentId", exam0.examId, ConfigurationCollection.grades(1))
  var exam0ResultEntry2: ExamResultEntry = ExamResultEntry(student2.username + "enrollmentId", exam0.examId, ConfigurationCollection.grades(2))

  var exam1ResultEntry0: ExamResultEntry = ExamResultEntry(student0.username + "enrollmentId", exam1.examId, ConfigurationCollection.grades.head)
  var exam1ResultEntry1: ExamResultEntry = ExamResultEntry(student1.username + "enrollmentId", exam1.examId, ConfigurationCollection.grades(1))
  var exam1ResultEntry2: ExamResultEntry = ExamResultEntry(student2.username + "enrollmentId", exam1.examId, ConfigurationCollection.grades(2))

  var exam3ResultEntry0: ExamResultEntry = ExamResultEntry(student0.username + "enrollmentId", exam3.examId, ConfigurationCollection.grades.head)
  var exam3ResultEntry1: ExamResultEntry = ExamResultEntry(student1.username + "enrollmentId", exam3.examId, ConfigurationCollection.grades(1))
  var exam3ResultEntry2: ExamResultEntry = ExamResultEntry(student2.username + "enrollmentId", exam3.examId, ConfigurationCollection.grades(2))
```

File: product_code/exam_result_service/impl/src/main/scala/de/upb/cs/uc4/examresult/impl/ExamResultServiceImpl.scala [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+                          throw UC4Exception.OwnerMismatch
+                        }
+                    }
+                }
+              }
+            case AuthenticationRole.Student if username.isEmpty || username.get.trim != authUser =>
+              throw UC4Exception.OwnerMismatch
+            case AuthenticationRole.Student =>
+              Future.successful(Done)
+            case _ => throw UC4Exception.InternalServerError("Error checking authRole", s"Role is not one of Student,Lecturer or Admin but instead ${role}")
```
Comment (dhelmert) [inline product_code/exam_result_service/impl/src/main/scala/de/upb/cs/uc4/examresult/impl/ExamResultServiceImpl.scala]:
```suggestion
            case _ => throw UC4Exception.InternalServerError("Error checking authRole", s"Role is not one of Student, Lecturer or Admin, but instead ${role}")
```

File: product_code/exam_result_service/impl/src/test/scala/de/upb/cs/uc4/examresult/impl/ExamResultServiceSpec.scala [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+    }
+    "not get the exam results of another student" in {
+      prepareSeq(allExamResults)
+      client.getExamResults(Some(student0.username), None).handleRequestHeader(addAuthorizationHeader(student1.username)).invoke().failed.map {
+        answer =>
+          answer.asInstanceOf[UC4Exception].possibleErrorResponse.`type` should ===(ErrorType.OwnerMismatch)
+      }
+    }
+
+    "get all exam results from an exam hold by the lecture himself" in {
```
Comment (dhelmert) [inline product_code/exam_result_service/impl/src/test/scala/de/upb/cs/uc4/examresult/impl/ExamResultServiceSpec.scala]:
```suggestion
    "get all exam results from an exam held by the lecturer themself" in {
```

File: product_code/exam_result_service/impl/src/test/scala/de/upb/cs/uc4/examresult/impl/ExamResultServiceSpec.scala [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    "get all exam results" in {
+      prepareSeq(allExamResults)
+      client.getExamResults(None, None).handleRequestHeader(addAuthorizationHeader()).invoke().map {
+        examResultsEntries => examResultsEntries should contain theSameElementsAs allExamResults
+      }
+    }
+
+    "get a students exam results as the student himself" in {
```
Comment (dhelmert) [inline product_code/exam_result_service/impl/src/test/scala/de/upb/cs/uc4/examresult/impl/ExamResultServiceSpec.scala]:
```suggestion
    "get a students exam results as the student themself" in {
```

File: product_code/exam_service/impl/src/test/scala/de/upb/cs/uc4/examresult/impl/ExamServiceSpec.scala [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+
+    "not get proposal add exam with an invalid moduleId" in {
+      val invalidExamModified = examOfLecturer0Modified.copy(moduleId = "NonExistingmoduleId")
+      client.getProposalAddExam.handleRequestHeader(addAuthorizationHeader(lecturer0.username)).invoke(invalidExamModified).failed.map {
+        answer =>
+          answer.asInstanceOf[UC4Exception].possibleErrorResponse.asInstanceOf[DetailedError].invalidParams.map(_.name) should
+            contain theSameElementsAs Seq("moduleId")
+      }
+    }
+    "not get proposal add exam for a non existing course" in {
```
Comment (dhelmert) [inline product_code/exam_service/impl/src/test/scala/de/upb/cs/uc4/examresult/impl/ExamServiceSpec.scala]:
```suggestion
    "not get proposal add exam for a non-existing course" in {
```

File: product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala [hunk 19]
```diff
@@ -131,26 +212,235 @@ class AdmissionServiceSpec extends AsyncWordSpec
...
+      }
+    }
+
+    " get all own course admissions as a student" in {
+      prepareSeq(allCourseAdmissions)(allExamAdmissions)
+      client.getCourseAdmissions(Some(student0.username), None, None).handleRequestHeader(addAuthorizationHeader(student0.username)).invoke().map {
+        courseAdmissions => courseAdmissions should contain theSameElementsAs Seq(defaultCourseAdmission0, defaultCourseAdmission1)
+      }
+    }
+    "get a course admissions from an own course as lecturer" in {
```
Comment (dhelmert) [inline product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala:229]:
```suggestion
    "get a course admissions to a course as the course's lecturer" in {
```

File: product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala [hunk 20]
```diff
@@ -131,26 +212,235 @@ class AdmissionServiceSpec extends AsyncWordSpec
...
+    }
+
+    "get all course admissions with specified moduleId as an admin" in {
+      prepareSeq(allCourseAdmissions)(allExamAdmissions)
+      client.getCourseAdmissions(None, None, Some(defaultCourse0.moduleIds.head)).handleRequestHeader(addAuthorizationHeader()).invoke().map {
+        courseAdmissions => courseAdmissions should contain theSameElementsAs Seq(defaultCourseAdmission0, defaultCourseAdmission2)
       }
     }
 
+    "get all course admissions with specified username,moduleId and courseID as an student" in {
```
Comment (dhelmert) [inline product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala:283]:
```suggestion
    "get all course admissions with specified username, moduleId and courseID as a student" in {
```

File: product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala [hunk 21]
```diff
@@ -131,26 +212,235 @@ class AdmissionServiceSpec extends AsyncWordSpec
...
+
+    //GET - Exam
+    "get all exam admissions as an admin" in {
+      prepareSeq(allCourseAdmissions)(allExamAdmissions)
+      client.getExamAdmissions(None, None, None).handleRequestHeader(addAuthorizationHeader()).invoke().map {
+        examAdmissions => examAdmissions should contain theSameElementsAs allExamAdmissions
+      }
+    }
+
+    " get all own exam admissions as a student" in {
```
Comment (dhelmert) [inline product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala:298]:
```suggestion
    "get all own exam admissions as a student" in {
```

File: product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala [hunk 22]
```diff
@@ -131,26 +212,235 @@ class AdmissionServiceSpec extends AsyncWordSpec
...
+
+    "not get exam admissions as another student" in {
+      prepareSeq(allCourseAdmissions)(allExamAdmissions)
+      client.getExamAdmissions(Some(student1.username), None, Some("RandomExamId")).handleRequestHeader(addAuthorizationHeader(student0.username)).invoke().failed.map {
+        answer =>
+          answer.asInstanceOf[UC4Exception].possibleErrorResponse.`type` should ===(ErrorType.OwnerMismatch)
+      }
+    }
+
+    " get all exam admissions from an exam as the lecturer which holds the exam" in {
```
Comment (dhelmert) [inline product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala:329]:
```suggestion
    "get all exam admissions from an exam as the lecturer which holds the exam" in {
```

File: product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala [hunk 23]
```diff
@@ -131,26 +212,235 @@ class AdmissionServiceSpec extends AsyncWordSpec
...
+      prepareSeq(allCourseAdmissions)(allExamAdmissions)
+      client.getExamAdmissions(None, None, Some(defaultExam0.examId)).handleRequestHeader(addAuthorizationHeader(lecturer0.username)).invoke().map {
+        examAdmissions => examAdmissions should contain theSameElementsAs Seq(defaultExamAdmission0, defaultExamAdmission1)
+      }
+    }
+
+    "not get a exam admission from an exam of another lecturer" in {
```
Comment (dhelmert) [inline product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala]:
```suggestion
    "not get an exam admission from an exam of another lecturer" in {
```

File: product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala [hunk 24]
```diff
@@ -131,26 +212,235 @@ class AdmissionServiceSpec extends AsyncWordSpec
...
+
+    "not get an exam admission if you are not enrolled" in {
+      prepareSeq(allCourseAdmissions)(allExamAdmissions)
+      client.getExamAdmissions(Some(student0.username + "NotEnrolledYet"), None, None).handleRequestHeader(addAuthorizationHeader(student0.username + "NotEnrolledYet")).invoke().failed.map {
+        answer =>
+          answer.asInstanceOf[UC4Exception].possibleErrorResponse.`type` should ===(ErrorType.KeyNotFound)
+      }
+    }
+
+    "not get a exam admission with an non-existent exam id" in {
```
Comment (dhelmert) [inline product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala]:
```suggestion
    "not get an exam admission with an non-existent exam id" in {
```

File: product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala [hunk 25]
```diff
@@ -131,26 +212,235 @@ class AdmissionServiceSpec extends AsyncWordSpec
...
+    }
+    "get the examAdmission with specified username, admissionIds, and examIds as a student" in {
+      prepareSeq(allCourseAdmissions)(allExamAdmissions)
+      client.getExamAdmissions(Some(student0.username), Some(defaultExamAdmission0.admissionId + "," + defaultExamAdmission2.admissionId), Some(defaultExam0.examId)).handleRequestHeader(addAuthorizationHeader(student0.username)).invoke().map {
+        examAdmissions => examAdmissions should contain theSameElementsAs Seq(defaultExamAdmission0)
+      }
+    }
+
+    //POST - negative generalAdd
+    "not get a proposal course admission with non empty enrollmentId,admissionId or timestamp" in {
```
Comment (dhelmert) [inline product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala]:
```suggestion
    "not get a proposal course admission with non empty enrollmentId, admissionId or timestamp" in {
```

File: product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala [hunk 26]
```diff
@@ -131,26 +212,235 @@ class AdmissionServiceSpec extends AsyncWordSpec
...
+      prepareSeq(allCourseAdmissions)(allExamAdmissions)
+      // here all three fields are non empty but in return we also get three validation errors
+      client.getProposalAddAdmission.handleRequestHeader(addAuthorizationHeader(student0.username)).invoke(defaultCourseAdmission0.copy(enrollmentId = "nonEmpty", admissionId = "nonEmpty", timestamp = "nonEmpty")).failed.map {
+        answer =>
+          answer.asInstanceOf[UC4Exception].possibleErrorResponse.asInstanceOf[DetailedError].invalidParams.map(_.name) should
+            contain theSameElementsAs Seq("admissionId", "enrollmentId", "timestamp")
+      }
+    }
+    "not get a proposal exam admission with non empty enrollmentId,admissionId or timestamp" in {
```
Comment (dhelmert) [inline product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala]:
```suggestion
    "not get a proposal exam admission with non empty enrollmentId, admissionId or timestamp" in {
```

File: product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala [hunk 27]
```diff
@@ -131,26 +212,235 @@ class AdmissionServiceSpec extends AsyncWordSpec
...
-      client.getProposalAddCourseAdmission.handleRequestHeader(addAuthorizationHeader(student0.username)).invoke(defaultAdmission).map {
+      client.getProposalAddAdmission.handleRequestHeader(addAuthorizationHeader(student0.username)).invoke(defaultCourseAdmission0.copy(enrollmentId = "", admissionId = "")).map {
         unsignedProposalJson =>
           asString(unsignedProposalJson.unsignedProposal).fromJson[CourseAdmission]
-            .copy(timestamp = "") should ===(defaultAdmission.copy(enrollmentId = certificate.get(student0.username).enrollmentId))
+            .copy(timestamp = "") should ===(defaultCourseAdmission0.copy(admissionId = "", enrollmentId = certificate.get(student0.username).enrollmentId))
+      }
+    }
+
+    "not get proposal add course admission " in {
```
Comment (dhelmert) [inline product_code/admission_service/impl/src/test/scala/de/upb/cs/uc4/admission/impl/AdmissionServiceSpec.scala]:
```suggestion
    "not get proposal add course admission" in {
```

**Review by dhelmert — APPROVED** [review]

**Pull request merged**